sa/rate_limits.go

```diff
@@ -0,0 +1,101 @@
+package sa
+
+import (
+	"database/sql"
+	"strings"
+	"time"
+
+	"github.com/letsencrypt/boulder/features"
+	"github.com/weppos/publicsuffix-go/publicsuffix"
+	"golang.org/x/net/context"
+)
+
+// baseDomain returns the eTLD+1 of a domain name for the purpose of rate
+// limiting. For a domain name that is itself an eTLD, it returns its input.
+func baseDomain(name string) string {
+	eTLDPlusOne, err := publicsuffix.Domain(name)
+	if err != nil {
+		// publicsuffix.Domain will return an error if the input name is itself a
+		// public suffix. In that case we use the input name as the key for rate
+		// limiting. Since all of its subdomains will have separate keys for rate
+		// limiting (e.g. "foo.bar.publicsuffix.com" will have
+		// "bar.publicsuffix.com", this means that domains exactly equal to a
+		// public suffix get their own rate limit bucket. This is important
+		// because otherwise they might be perpetually unable to issue, assuming
+		// the rate of issuance from their subdomains was high enough.
+		return name
+	}
+	return eTLDPlusOne
+}
+
+// addCertificatesPerName adds 1 to the rate limit count for the provided domains,
+// in a specific time bucket. It must be executed in a transaction.
```

Comment:
Probably should either say this bucket is an hour, or change the arg name below to be more generic.